src/scheduler/work_bucket.rs

```diff
@@ -1,147 +1,200 @@
             monitor,
             can_open: None,
+            group,
         }
     }
+
+    #[inline(always)]
     fn notify_one_worker(&self) {
-        let _guard = self.monitor.0.lock().unwrap();
-        self.monitor.1.notify_one()
+        // If the bucket is not activated, don't notify anyone.
+        if !self.is_activated() {
+            return;
+        }
+        // Notify one if there're any parked workers.
+        if self.group.parked_workers() > 0 {
+            let _guard = self.monitor.0.lock().unwrap();
+            self.monitor.1.notify_one()
+        }
     }
-    fn notify_all_workers(&self) {
-        let _guard = self.monitor.0.lock().unwrap();
-        self.monitor.1.notify_all()
+
+    #[inline(always)]
+    pub fn notify_all_workers(&self) {
+        // If the bucket is not activated, don't notify anyone.
+        if !self.is_activated() {
+            return;
+        }
+        // Notify all if there're any parked workers.
+        if self.group.parked_workers() > 0 {
+            let _guard = self.monitor.0.lock().unwrap();
+            self.monitor.1.notify_all()
+        }
     }
+
+    #[inline(always)]
     pub fn is_activated(&self) -> bool {
-        self.active.load(Ordering::SeqCst)
+        self.active.load(Ordering::Relaxed)
     }
+
     /// Enable the bucket
     pub fn activate(&self) {
-        self.active.store(true, Ordering::SeqCst);
+        self.active.store(true, Ordering::Relaxed);
     }
+
     /// Test if the bucket is drained
+    #[inline(always)]
     pub fn is_empty(&self) -> bool {
-        self.queue.read().len() == 0
+        self.queue.is_empty()
+            && self
+                .prioritized_queue
+                .as_ref()
+                .map(|q| q.is_empty())
+                .unwrap_or(true)
     }
+
+    #[inline(always)]
     pub fn is_drained(&self) -> bool {
         self.is_activated() && self.is_empty()
     }
+
     /// Disable the bucket
     pub fn deactivate(&self) {
-        debug_assert!(
-            self.queue.read().is_empty(),
-            "Bucket not drained before close"
-        );
-        self.active.store(false, Ordering::SeqCst);
-    }
-    /// Add a work packet to this bucket, with a given priority
-    pub fn add_with_priority(&self, priority: usize, work: Box<dyn GCWork<VM>>) {
-        self.queue
-            .write()
-            .push(PrioritizedWork::new(priority, work));
-        self.notify_one_worker(); // FIXME: Performance
-    }
-    /// Add a work packet to this bucket, with a default priority (1000)
+        debug_assert!(self.queue.is_empty(), "Bucket not drained before close");
+        self.active.store(false, Ordering::Relaxed);
+    }
+
+    /// Add a work packet to this bucket
+    /// Panic if this bucket cannot receive prioritized packets.
+    #[inline(always)]
+    pub fn add_prioritized(&self, work: Box<dyn GCWork<VM>>) {
+        self.prioritized_queue.as_ref().unwrap().push(work);
+        self.notify_one_worker();
```

Comment:
I thought about this a little bit more. Looks like option 1 above is a necessary step before implementing the design proposed at https://github.com/mmtk/mmtk-core/issues/546. It builds a bottleneck for adding work packets, and the design in https://github.com/mmtk/mmtk-core/issues/546 can utilize it to either redirect each packet to the main queue or temporarily store it in a bucket.

I will implement option 1 as a separate PR (this will likely to introduce many changes in mmtk-core and bindings), and have a try on https://github.com/mmtk/mmtk-core/issues/546.